Ad7192 dev

romandariana opened the pull request:
Update some features.
Add support for AD7194.

nunojsa commented:
It seems there are valid CI failures... Please address them

amiclaus commented on drivers/iio/adc/ad7192.c:
```diff
@@ -267,7 +267,7 @@ static struct ad7192_state *ad_sigma_delta_to_ad7192(struct ad_sigma_delta *sd)
 
 static int ad7192_set_channel(struct ad_sigma_delta *sd, unsigned int slot,
-	unsigned int channel)
+			      unsigned int channel)
```
nit: unrelated change

amiclaus commented on drivers/iio/adc/ad7192.c:
```diff
@@ -492,7 +492,7 @@ static ssize_t ad7192_set(struct device *dev,
 
 static void ad7192_get_available_filter_freq(struct ad7192_state *st,
-						    int *freq)
+					     int *freq)
```
ditto

amiclaus commented on arch/arm/boot/dts/zynq-coraz7s-ad7190.dts:
```diff
@@ -0,0 +1,45 @@
+        avdd-supply = <&avdd>;
+        dvdd-supply = <&dvdd>;
+    };
+};
```
nit: new line

amiclaus commented on arch/arm/boot/dts/zynq-coraz7s-ad7190.dts:
```diff
@@ -0,0 +1,45 @@
+ *
+ * hdl_project: <ad719x_asdz/coraz7s>
+ *
+ * Copyright (C) 2022 Analog Devices, Inc.
```
2023

amiclaus commented on arch/arm/boot/dts/zynq-coraz7s-ad7193.dts:
```diff
@@ -0,0 +1,45 @@
+ *
+ * hdl_project: <ad719x_asdz/coraz7s>
+ *
+ * Copyright (C) 2022 Analog Devices, Inc.
```
2023

amiclaus commented on arch/arm/boot/dts/zynq-coraz7s-ad7193.dts:
```diff
@@ -0,0 +1,45 @@
+        avdd-supply = <&avdd>;
+        dvdd-supply = <&dvdd>;
+    };
+};
```
nit: new line

amiclaus commented on arch/arm/boot/dts/zynq-coraz7s-ad7195.dts:
```diff
@@ -0,0 +1,45 @@
+        avdd-supply = <&avdd>;
+        dvdd-supply = <&dvdd>;
+    };
+};
```
ditto

amiclaus commented on drivers/iio/adc/Kconfig:
```diff
@@ -100,12 +100,12 @@ config AD7173
 
 config AD7192
-	tristate "Analog Devices AD7190 AD7192 AD7193 AD7195 ADC driver"
+	tristate "Analog Devices AD7190 AD7192 AD7193 AD7194 AD7195 ADC driver"
```
i think this commit should be squashed with the one adding the support for ad7194 in the driver.

amiclaus commented on arch/arm/boot/dts/zynq-coraz7s-ad7194.dts:
```diff
@@ -0,0 +1,56 @@
+                };
+        };
+    };
+};
```
ditto

amiclaus commented on arch/arm/boot/dts/zynq-coraz7s-ad7194.dts:
```diff
@@ -0,0 +1,56 @@
+ *
+ * hdl_project: <ad719x_asdz/coraz7s>
+ *
+ * Copyright (C) 2022 Analog Devices, Inc.
```
2023

amiclaus reviewed (commented):
some comments on my side

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -128,10 +128,26 @@
+					/* Differential Channel Address Mask */
+#define AD7194_CH_TEMP			0x100	/* Temp sensor */
+#define AD7194_CH_AIN_MASK(x)	(((x) - 1) && 0xF)
+#define AD7194_CH_AIN(x)		((AD7194_CH_AIN_MASK(x) << 4) | 0x400)
```
nit: could align this with the previous line...

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -674,6 +692,17 @@ static const struct attribute_group ad7193_attribute_group = {
+	&iio_dev_attr_filter_low_pass_3db_frequency_available.dev_attr.attr,
+	&iio_dev_attr_fast_settling_average_factor.dev_attr.attr,
+	&iio_dev_attr_fast_settling_average_factor_available.dev_attr.attr,
+	NULL
```
This new attributes have to be documented in here:

https://elixir.bootlin.com/linux/v6.2-rc7/source/Documentation/ABI/testing/sysfs-bus-iio-adc-ad7192

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -1009,21 +1059,116 @@ static const struct ad7192_chip_info ad7192_chip_info_tbl[] = {
+	if (num_channels > AD7194_CH_DIFF_NR) {
+		dev_err(indio_dev->dev.parent, "Differential channels number out of range.\n");
+		return -EINVAL;
+	}
```
what if channels_node is NULL?... Instead do:

```
if (!channels_node)
    return 0; // assuming is not a mandatory property
```

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -1009,21 +1059,116 @@ static const struct ad7192_chip_info ad7192_chip_info_tbl[] = {
+		num_channels = of_get_available_child_count(channels_node);
+
+	if (num_channels > AD7194_CH_DIFF_NR) {
+		dev_err(indio_dev->dev.parent, "Differential channels number out of range.\n");
```
initialize a `struct device *dev = indio_dev->dev.parent` and use it instead...

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -1009,21 +1059,116 @@ static const struct ad7192_chip_info ad7192_chip_info_tbl[] = {
+	indio_dev->num_channels = num_channels;
+
+	if (!channels_node)
+		return 0;
```
This should be right after getting (tying) the child node...

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -1009,21 +1059,116 @@ static const struct ad7192_chip_info ad7192_chip_info_tbl[] = {
+		ret = of_property_read_u32_array(child, "reg", ain, 2);
+		if (ret) {
+			of_node_put(channels_node);
+			of_node_put(child);
```
nit: maybe an `out_error` label would good...

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -1009,21 +1059,116 @@ static const struct ad7192_chip_info ad7192_chip_info_tbl[] = {
+
+		scan_index++;
+		ad7194_channels++;
+	}
```
nit: extra line

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -552,6 +662,18 @@ static const struct attribute_group ad7192_attribute_group = {
+
+static const struct attribute_group ad7193_attribute_group = {
+	.attrs = ad7193_attributes,
+};
```
ditto for the ABI documentation...

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -523,17 +608,38 @@ static ssize_t ad7192_show_filter_avail(struct device *dev,
+				 st->avg_avail[i]);
+
+	buf[len - 1] = '\n';
+
```
It would be nice to refactor a bit these places to use `sysfs_emit()`. You would, naturally, need a precursor patch doing the change for the "old" code so that when introducing this patch, you would disrupt the driver "style".

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -491,22 +529,69 @@ static ssize_t ad7192_set(struct device *dev,
+	u8 avg_factor;
+	u8 avg_factor_selected;
+
+	avg_factor_selected = (st->mode & AD7192_MODE_AVG_MASK) >> 16;
```
`FIELD_GET()`... Not sure if the driver is already using these macros though

romandariana replied:
The driver is not using FIELD_GET(). Should I add a commit to update the driver to use it everywhere or leave it like this?

pcercuei replied:
Yes, good idea.

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -491,22 +529,69 @@ static ssize_t ad7192_set(struct device *dev,
+	avg_factor_selected = (st->mode & AD7192_MODE_AVG_MASK) >> 16;
+
+	if (!(avg_factor_selected) &&
+	    !(chop_enabled))
```
I think this can be one liner?

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -491,22 +529,69 @@ static ssize_t ad7192_set(struct device *dev,
+	if (sinc3_enabled)
+		sinc_factor = AD7192_SYNC3_FILTER;
+
+	avg_factor = st->avg_avail[avg_factor_selected];
```
you just check `avg_factor_selected` for non 0... Are we sure we cannot go out of boundaries in here?

romandariana replied:
#define AD7192_MODE_AVG_MASK	(0x3 << 16)
...
u8 avg_factor_selected;
avg_factor_selected = (st->mode & AD7192_MODE_AVG_MASK) >> 16;
I believe avg_factor_selected can only be 0, 1, 2 or 3.

nunojsa reviewed (commented):
Also, make sure to send all of these patches upstream... We'll merge it after it gets accepted in there

romandariana commented on drivers/iio/adc/ad7192.c:
```diff
@@ -90,6 +89,7 @@
 /* Configuration Register Bit Designations (AD7192_REG_CONF) */
 
 #define AD7192_CONF_CHOP	BIT(23) /* CHOP enable */
+#define AD7192_CONF_ACX		BIT(22) /* AC excitation enable(AD7195 only) */
```
so it would be better to go with the old name, even though the old name includes the wrong register?

nunojsa replied:
Continuing this discussion... I guess the name change makes sense but you could/should justify it in the commit message.

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -836,18 +836,14 @@ static const struct iio_info ad7195_info = {
 	AD719x_DIFF_CHANNEL(1, 3, 4, AD7192_CH_AIN3P_AIN4M),
 	AD719x_TEMP_CHANNEL(2, AD7192_CH_TEMP),
-	AD719x_SHORTED_CHANNEL(3, 2, AD7192_CH_AIN2P_AIN2M),
+	AD719x_DIFF_CHANNEL(3, 2, 2, AD7192_CH_AIN2P_AIN2M),
```
> So maybe it is OK to fix it?

That's a good question... I guess the only thing to do is to ask Jonathan in the mailing list about this mess :sweat_smile:

romandariana commented:
V1 -> V2
- Corrected mistakes.
- Changed AD7194 differential channels configuration function to be in tone with models I found upstream.
- Didn't change libiio related problem for now(I will ask in mailing list).

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -1019,21 +1071,105 @@ static const struct ad7192_chip_info ad7192_chip_info_tbl[] = {
+	for_each_available_child_of_node(np, child) {
+		ret = of_property_read_u32_array(child, "diff-channels", ain, 2);
+		if (ret)
+			goto out_error;
```
if they are really required properties, you need to say that in the bindings...

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -976,6 +1016,18 @@ static const struct iio_info ad7195_info = {
+
+static const struct iio_chan_spec ad7194_channel_timestamp_template =
+	IIO_CHAN_SOFT_TIMESTAMP(0);
+
```
no need to change... but instead of this you can have a macro that you could then use to assign/fill the channels

pcercuei commented on drivers/iio/adc/ad7192.c:
```diff
@@ -861,7 +857,7 @@ static const struct iio_chan_spec ad7193_channels[] = {
 	AD719x_DIFF_CHANNEL(3, 7, 8, AD7193_CH_AIN7P_AIN8M),
 	AD719x_TEMP_CHANNEL(4, AD7193_CH_TEMP),
-	AD719x_SHORTED_CHANNEL(5, 2, AD7193_CH_AIN2P_AIN2M),
+	AD719x_DIFF_CHANNEL(5, 2, 2, AD7193_CH_AIN2P_AIN2M),
```
That was my suggestion for the fix - as a shorted channel is basically a differential with itself. But that was before I realized that the bug is already upstream and not something added by the PR.

romandariana commented on Documentation/devicetree/bindings/iio/adc/adi,ad7192.yaml:
```diff
@@ -94,30 +102,86 @@ required:
+  "^channel@([0-9a-f])$":
+    type: object
+    $ref: adc.yaml
+    unevaluatedProperties: false
```
@nunojsa this is the order I found in multiple drivers upstream, is it not correct?

romandariana commented:
V2 -> V3

- Divided the "add fast settling support" commit into 2 commits
- Removed the "fix shorted channel" commit which was sent upstream already

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -503,22 +501,61 @@ static ssize_t ad7192_set(struct device *dev,
+{
+	unsigned int f_order;
+
+	f_order = ad7192_get_f_order(st);
```
just do this above...

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -503,22 +501,61 @@ static ssize_t ad7192_set(struct device *dev,
+
+	f_order = ad7192_get_f_order(st);
+
+	return DIV_ROUND_CLOSEST(st->fclk, f_order * AD7192_MODE_RATE(st->mode));
```
can `f_order` ever be 0?

romandariana replied:
no, the f_order computing functions will return a positive value

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -485,13 +510,34 @@ static ssize_t ad7192_set(struct device *dev,
+			if (val_avg_factor == st->avg_avail[i]) {
+				st->mode &= ~AD7192_MODE_AVG_MASK;
+				st->mode |= AD7192_MODE_AVG(i);
+				ad_sd_write_reg(&st->sd, AD7192_REG_MODE, 3, st->mode);
```
is `ad_sd_write_reg()` returning an error code? If so, you should check it...

nunojsa commented on drivers/iio/adc/ad7192.c:
```diff
@@ -1026,21 +1078,104 @@ static const struct ad7192_chip_info ad7192_chip_info_tbl[] = {
+	u32 ain[2];
+	int ret = 0;
+
+	num_channels = of_get_available_child_count(np);
```
Just a comment.... I see the driver already uses `of_device_get_match_data()` but, AFAICS, is the only OF specific bit used. So you might be asked to convert it to the device API and also use that in here...

nunojsa reviewed (commented):
Just some comments from my side... Just consider my comments and take this upstream :)

One other note though:

> Removed the "fix shorted channel" commit which was sent upstream already

You should then cherry-pick the patch accepted upstream and either update this PULL or create a new one with it.

nunojsa commented:
@romandariana, what's the status of this work? Do we want to keep the PULL open? Should I close it?
